rust/frontend/src/ndarray.rs

```diff
@@ -0,0 +1,359 @@
+                v.set_len(sz);
+                v
+            })
+        }
+    }
+
+    #[inline]
+    pub fn is_contiguous(&self) -> bool {
+        self.strides().is_none()
+    }
+
+    pub fn byte_offset(&self) -> isize {
+        unsafe { (*self.handle).byte_offset as isize }
+    }
+
+    /// Flattens the NDArray to a Vec of the same type in cpu.
+    ///
+    /// ## Example
+    ///
+    /// ```
+    /// let mut shape = vec![4];
+    /// let mut data = vec![1i32, 2, 3, 4];
+    /// let ctx = TVMContext::cpu(0);
+    /// let mut ndarray = empty(&mut shape, ctx, TVMType::from("int"));
+    /// ndarray.copy_from_buffer(&mut data);
+    /// assert_eq!(ndarray.shape(), Some(shape));
+    /// assert_eq!(ndarray.to_vec::<i32>().unwrap(), data);
+    /// ```
+    pub fn to_vec<T>(&self) -> Result<Vec<T>> {
+        if self.shape().is_none() {
+            return Err(Error::EmptyArray);
+        }
+        let earr = empty(&mut self.shape().unwrap(), TVMContext::cpu(0), self.dtype());
+        let target = self.copy_to_ndarray(earr).unwrap();
+        let arr = unsafe { *(target.handle) };
+        let sz = self.size().unwrap() as usize;
+        let mut v: Vec<T> = Vec::with_capacity(sz * mem::size_of::<T>());
+        unsafe {
+            v.as_mut_ptr()
+                .copy_from_nonoverlapping(arr.data as *const T, sz);
+            v.set_len(sz);
+        }
+        Ok(v)
+    }
+
+    /// Converts the NDArray to [`TVMByteArray`].
+    pub fn to_bytearray(&self) -> Result<TVMByteArray> {
+        let v = self.to_vec::<u8>()?;
+        Ok(TVMByteArray::from(&v))
+    }
+
+    /// Creates an NDArray from a mutable buffer of types i32, u32 or f32 in cpu.
+    /// ## Example
+    ///
+    /// ```
+    /// let mut shape = vec![2];
+    /// let mut data = vec![1f32, 2];
+    /// let ctx = TVMContext::gpu(0);
+    /// let mut ndarray = empty(&mut shape, ctx, TVMType::from("int"));
+    /// ndarray.copy_from_buffer(&mut data);
+    /// ```
+    pub fn copy_from_buffer<T: Num32>(&mut self, data: &mut [T]) {
+        check_call!(ts::TVMArrayCopyFromBytes(
+            self.handle,
+            data.as_ptr() as *mut _,
+            data.len() * mem::size_of::<T>()
+        ));
+    }
+
+    /// Copies the NDArray to another target NDArray.
+    pub fn copy_to_ndarray(&self, target: NDArray) -> Result<NDArray> {
+        if self.dtype() != target.dtype() {
+            return Err(Error::TypeMismatch {
+                expected: self.dtype().to_string(),
+                found: target.dtype().to_string(),
+            });
+        }
+        check_call!(ts::TVMArrayCopyFromTo(
+            self.handle,
+            target.handle,
+            ptr::null_mut() as ts::TVMStreamHandle
+        ));
+        Ok(target)
+    }
+
+    /// Copies the NDArray to a target context.
+    pub fn copy_to_ctx(&self, target: &TVMContext) -> Result<NDArray> {
+        let tmp = empty(&mut self.shape().unwrap(), target.clone(), self.dtype());
```

Comment:
using `unwrap` in libraries is generally a bad idea if it's possible to panic. please use `?` since you're already returning a `Result`

Reply:
right! but shape is an Option and what would mean to propagation [NoneError](https://doc.rust-lang.org/std/option/struct.NoneError.html) then.

Reply:
sure, that's fine. just do `ok_or(...)?` then

Reply:
I added `NoneError` to error-chain and it works like a charm.

Reply:
nice! might I suggest a more descriptive name like `MissingShapeError` or something to that effect?

Reply:
Sure, makes sense :+1: